### PR TITLE
[WIP] add protocol features and minor fixes

### PR DESCRIPTION
**Pull request by zuntrax**
current commands:
* `message`
* `set_video_a`
* `get_video_a`
* `set_video_b`
* `get_video_b`
* `set_audio`
* `get_audio`
* `set_composite_mode`
* `get_composite_mode`
* `set_stream_status`
* `get_stream_status`
* `get_config`
`get_config` gives a base64-encoded ini-dump of the current config

### CODE REVIEW

File: voctocore/lib/commands.py [hunk 1]
```diff
@@ -60,11 +87,38 @@ def set_composite_mode(self, composite_mode):
...
 		return True
 
-	def set_stream_blank(self, src_name_or_id):
-		src_id = self.decodeBlankerSourceName(src_name_or_id)
-		self.pipeline.streamblanker.setBlankSource(src_id)
-		return True
+	def get_composite_mode(self):
+		try:
+			mode = self.pipeline.vmix.getCompositeMode()
+			name = str(mode).split('.')[-1]
```
Comment (MaZderMind) [inline voctocore/lib/commands.py]: why is that split there? if it is required, shouldn't it go into vmix.getCompositeMode?

File: voctocore/lib/commands.py [hunk 2]
```diff
@@ -60,11 +87,38 @@ def set_composite_mode(self, composite_mode):
...
+		try:
+			if args[0] == "live":
+				self.pipeline.streamblanker.setBlankSource(None)
+				return True
+			elif args [0] == "blank":
+				src_id = self.decodeBlankerSourceName(args[1])
+				self.pipeline.streamblanker.setBlankSource(src_id)
+				return True
+			else:
+				return (False, "invalid arguments")
```
Comment (MaZderMind) [inline voctocore/lib/commands.py]: a more descriptive error, like "only a or b is alloes as an argument to …" would be great

File: voctocore/lib/commands.py [hunk 3]
```diff
@@ -60,11 +87,38 @@ def set_composite_mode(self, composite_mode):
...
+		except KeyError as e:
+			raise KeyError("invalid arguments")
```
Comment (MaZderMind) [inline voctocore/lib/commands.py]: where does this exception come from? is it from decodeBlankerSourceName? shouldn't it then be insise the elif? also, a ore descriptive explanation of the error would be nice

**Comment by MaZderMind** [comment]
Looks pretty good! I commented three pieces above. After you'll fixed them I'll merge it!

**Comment by MaZderMind** [comment]
One thing though: you mention get_config in the PR's text but it is not in the code. Also - why did you opt for base64-ini – which is a pretty uncommon format.

File: voctocore/lib/commands.py [hunk 4]
```diff
@@ -60,11 +87,38 @@ def set_composite_mode(self, composite_mode):
...
+			return (True, "live")
+
+		name = self.encodeBlankerSourceName(self.pipeline.streamblanker.blankSource)
+		return (True, "blank " + name)
+
+	def get_config(self):
+		dump = StringIO()
+		Config.write(dump)
+		dump.seek(0)
+		return (True, base64.b64encode(dump.read().encode()).decode())
```
Comment (MaZderMind) [inline voctocore/lib/commands.py]: better build a dict here and use json as transport format: `{k: dict(v) for k, v in  dict(Config).items()}`
you can use `read_dict` on the gui to overlay the server config over the gui config.

File: voctocore/lib/commands.py (hunk 1 above)
Comment (MaZderMind) [inline voctocore/lib/commands.py]: better use better use `mode.name` here

File: voctocore/lib/commands.py (hunk 3 above)
Comment (zuntrax) [inline voctocore/lib/commands.py]: if too few args are given, args[0] or args[1] might throw this exception.

**Pull request merged**